shared/typeflow/codeql/typeflow/internal/TypeFlowImpl.qll

```diff
@@ -0,0 +1,453 @@
+private import codeql.typeflow.TypeFlow
+private import codeql.util.Location
+private import codeql.util.Unit
+
+module TypeFlow<LocationSig Location, TypeFlowInput<Location> I> {
+  private import I
+
+  /** Holds if `null` is the only value that flows to `n`. */
+  predicate isNull(TypeFlowNode n) {
+    isNullValue(n)
+    or
+    exists(TypeFlowNode mid | isNull(mid) and step(mid, n))
```

Comment:
This appears to be missing the `forex` case over `joinStep`.

Reply:
Hmm... I thought you were right here, but it's actually included in the Java instantiation [here](https://github.com/github/codeql/pull/15728/commits/6a65c46b2e7fe3888a2c3c025751488eb848991b#diff-99689bd4f8f452c2de25ff84c624f610de94ae4cae5c6158aa5a88024fa4fb85R137) since it also needs to be joined with `not exists(n.asField())`. That's not super pretty, I know 😬

As an alternative, we can have have some kind of exclusion predicate in the interface, and have:
```ql
forex(TypeFlowNode mid | joinStep0(mid, n) | Make<J::Location, Input>::isNull(mid)) and
not excluded(n)
```
Then Java can implement `excluded(n)` as `exists(n.asField())`

Reply:
I've done the above in 1acbb84444ba40ca106f6bf529a877f39d61ed8a. Hopefully that should make it clear that the rewrite preserves the meaning of `isNull` 🤞